test/response_test.rb

```diff
@@ -34,380 +31,913 @@ class RubySamlTest < Minitest::Test
+      response.settings = settings
+      assert_raises(OneLogin::RubySaml::ValidationError, "There is an EncryptedAttribute in the Response and this SP not support them") do
+        response.validate!
       end
+    end
 
-      it "return true when using certificate instead of fingerprint" do
-        response = OneLogin::RubySaml::Response.new(response_document_4)
-        response.stubs(:conditions).returns(nil)
-        settings = OneLogin::RubySaml::Settings.new
-        response.settings = settings
-        settings.idp_cert = signature_1
-        assert response.is_valid?
+    it "raise when there is no valid audience" do
+      response = OneLogin::RubySaml::Response.new(valid_signed_response)
+      response.settings = settings
+      response.settings.issuer = 'invalid'
+      assert_raises(OneLogin::RubySaml::ValidationError, "#{response.settings.issuer} is not a valid audience for this Response") do
+        response.validate!
       end
+    end
 
-      it "not allow signature wrapping attack" do
-        response = OneLogin::RubySaml::Response.new(response_document_4)
-        response.stubs(:conditions).returns(nil)
-        settings = OneLogin::RubySaml::Settings.new
-        settings.idp_cert_fingerprint = signature_fingerprint_1
-        response.settings = settings
-        assert response.is_valid?
-        assert_equal response.name_id, "test@onelogin.com"
+    it "raise when the destination od the SAML Response does not match the assertion consumer service url" do
+      response = OneLogin::RubySaml::Response.new(valid_signed_response)
+      response.settings = settings
+      response.settings.assertion_consumer_service_url = 'invalid_acs'
+      assert_raises(OneLogin::RubySaml::ValidationError, "The response was received at #{response.destination} instead of #{response.settings.assertion_consumer_service_url}") do
+        response.validate!
       end
+    end
 
-      it "support dynamic namespace resolution on signature elements" do
-        response = OneLogin::RubySaml::Response.new(fixture("no_signature_ns.xml"))
-        response.stubs(:conditions).returns(nil)
-        settings = OneLogin::RubySaml::Settings.new
-        response.settings = settings
-        settings.idp_cert_fingerprint = "28:74:9B:E8:1F:E8:10:9C:A8:7C:A9:C3:E3:C5:01:6C:92:1C:B4:BA"
-        XMLSecurity::SignedDocument.any_instance.expects(:validate_signature).returns(true)
-        assert response.validate!
+    it "raise when  the issuer of the Message does not match the IdP entityId" do
+      response = OneLogin::RubySaml::Response.new(response_invalid_issuer_message)
+      response.settings = settings
+      response.settings.idp_entity_id = 'http://idp.example.com/'
+      assert_raises(OneLogin::RubySaml::ValidationError, "Doesn't match the issuer, expected: <#{response.settings.idp_entity_id}>, but was: <http://invalid.issuer.example.com/>") do
+        response.send(:validate_issuer, false)
```

Comment:
I could call the public validate method, but other errors can take place instead of the method that I wanted to test. In order to avoid this conflicts on test and make it easier I'm testing each validation method,